Plain JS binder errors

sandersn opened the pull request:
Issue some errors from the binder in JS files that do not have checkJS explicitly turned on or off. These errors mirror runtime checks done by Javascript. The checker has many more errors that also cover runtime checks. I'll start on those next.

There are a few errors in the binder that don't have matching runtime errors: they *nearly* always indicate incorrect code, but not always. 

### A label is not allowed here

I decided to include "A label is not allowed here" because, even though it's not a runtime error on its own, if you use the label it will not be defined, which *is* a runtime error, and because the compiler error for the undefined label is so bad.

### An object literal cannot have multiple properties with the same name in strict mode

I decided not to include this error, since it's not a runtime error in ES2015 and above according to an SO answer I found. It's a useful error for TS, though; probably the error message should be changed to read "An object literal **should** not have multiple properties with the same name.", or else it should be restricted to ES5-only. I filed https://github.com/microsoft/TypeScript/issues/46815 to track that.

Addresses but does not fully fix #45349.

sandersn commented:
@mjbvz You might also have opinions about the appropriateness of these errors.

sandersn commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,21 @@ namespace ts {
+                }
 
-                return getMergedBindAndCheckDiagnostics(sourceFile, includeBindAndCheckDiagnostics, bindDiagnostics, checkDiagnostics, isCheckJs ? sourceFile.jsDocDiagnostics : undefined);
+                return getMergedBindAndCheckDiagnostics(sourceFile, includeBindAndCheckDiagnostics && !isPlainJs, bindDiagnostics, checkDiagnostics, isCheckJs ? sourceFile.jsDocDiagnostics : undefined);
```
this is needed to avoid bogus errors from `ts-expect-error` in plain JS files.

amcasey replied:
Consider adding a comment to that effect.

DanielRosenwasser commented:
> I decided to include "A label is not allowed here" because, even though it's not a runtime error on its own, if you use the label it will not be defined, which is a runtime error, and because the compiler error for the undefined label is so bad.

https://github.com/microsoft/TypeScript/issues/30408

> I decided not to include this error, since it's not a runtime error in ES2015 and above according to an SO answer I found. It's a useful error for TS, though; probably the error message should be changed to read "An object literal should not have multiple properties with the same name.", or else it should be restricted to ES5-only. I filed #46815 to track that.

My 2 cents is to just make it a suggestion diagnostic; don't go too crazy with when and how to report this in JS, or modifying how it surfaces in TS.

DanielRosenwasser commented:
@typescript-bot test this

amcasey commented on src/compiler/program.ts:
```diff
@@ -818,6 +818,25 @@ namespace ts {
+        Diagnostics.Identifier_expected_0_is_a_reserved_word_that_cannot_be_used_here.code,
+        Diagnostics.constructor_is_a_reserved_word.code,
+        Diagnostics.delete_cannot_be_called_on_an_identifier_in_strict_mode.code,
+        Diagnostics.Code_contained_in_a_class_is_evaluated_in_JavaScript_s_strict_mode_which_does_not_allow_this_use_of_0_For_more_information_see_https_Colon_Slash_Slashdeveloper_mozilla_org_Slashen_US_Slashdocs_SlashWeb_SlashJavaScript_SlashReference_SlashStrict_mode.code,
```
This is a hilarious property name.

sandersn replied:
Yeah, I'm not super happy about urls in error messages either.

DanielRosenwasser replied:
I had the honor of writing the identifier generator and you're welcome.

amcasey replied:
On the VS side, we tend to use aka.ms or fwlinks so that we can update them if the target ever changes.

@DanielRosenwasser No underscore after the protocol? 😛

amcasey commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,21 @@ namespace ts {
                 Debug.assert(!!sourceFile.bindDiagnostics);
 
-                const isCheckJs = isCheckJsEnabledForFile(sourceFile, options);
+                const isCheckJs = isCheckJsEnabledForFile(sourceFile, options) && (sourceFile.scriptKind === ScriptKind.JS || sourceFile.scriptKind === ScriptKind.JSX);
```
Pull out the duplicated script kind check?

amcasey commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,21 @@ namespace ts {
+                const isPlainJs = !sourceFile.checkJsDirective && options.checkJs === undefined && (sourceFile.scriptKind === ScriptKind.JS || sourceFile.scriptKind === ScriptKind.JSX);
                 const isTsNoCheck = !!sourceFile.checkJsDirective && sourceFile.checkJsDirective.enabled === false;
+
                 // By default, only type-check .ts, .tsx, 'Deferred' and 'External' files (external files are added by plugins)
```
Update comment?

amcasey reviewed (approved):
Seems sensible.

DanielRosenwasser commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,21 @@ namespace ts {
+                let bindDiagnostics: readonly Diagnostic[] = includeBindAndCheckDiagnostics ? sourceFile.bindDiagnostics : emptyArray;
+                let checkDiagnostics = includeBindAndCheckDiagnostics ? typeChecker.getDiagnostics(sourceFile, cancellationToken) : emptyArray;
+                if (isPlainJs) {
+                    bindDiagnostics = bindDiagnostics.filter(d => plainJSErrors.has(d.code));
+                    checkDiagnostics = checkDiagnostics.filter(d => plainJSErrors.has(d.code));
```
Small, but `filter` in `core.ts` avoids creating a new array if the elements are the same.

sandersn replied:
I made the change because I like to pretend JS is actually Scheme with bad syntax, but it's super unlikely that (1) the diagnostics array would be very large at the same time as (2) every single error was in the plainJSErrors set. This code is more likely to do the opposite: filter out every item of the array.

DanielRosenwasser commented:
@typescript-bot pack this

orta commented:
@typescript-bot pack this

sandersn commented:
I'm going to hold this until we have a chance to discuss this in the editor meeting tomorrow.

andrewbranch commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,25 @@ namespace ts {
 
-                const isCheckJs = isCheckJsEnabledForFile(sourceFile, options);
+                const isJs = sourceFile.scriptKind === ScriptKind.JS || sourceFile.scriptKind === ScriptKind.JSX;
+                const isCheckJs = isJs && isCheckJsEnabledForFile(sourceFile, options);
+                const isPlainJs = isJs && !sourceFile.checkJsDirective && options.checkJs === undefined;
```
What’s the difference between `isCheckJsEnabledForFile` the property checks you’re doing for `isPlainJs`?

sandersn replied:
I *almost* inlined that function to make this easier to read, except that it's used everywhere. `isCheckJsEnabledForFile` is true whenever checkJs is explicitly true, whether locally or globally. To put it in the most parallel form:

`sourceFile.checkJsDirective?.enabled || options.checkJs === true` vs
`sourceFile.checkJsDirective === undefined && options.checkJs === undefined`

andrewbranch replied:
Ah, so the third kind of JS is JS where `checkJs` is explicitly _disabled_?

sandersn replied:
That's right. `// @ts-nocheck` is the local form of that.

andrewbranch commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,25 @@ namespace ts {
+                // - check JS: .js files with either // ts-check or checkJs: true
+                // - external: files that are added by plugins
+                const includeBindAndCheckDiagnostics = !isTsNoCheck && (sourceFile.scriptKind === ScriptKind.TS || sourceFile.scriptKind === ScriptKind.TSX
+                        || sourceFile.scriptKind === ScriptKind.External || isPlainJs || isCheckJs || sourceFile.scriptKind === ScriptKind.Deferred);
```
> `isPlainJs || isCheckJs || ...`

Related to my other question, but what other kind of JS is there?

It took me a second to realize, it’s easier to think about this condition in terms of what it excludes, although it may or may not be easier to write it that way:

- `// @ts-nocheck` files
- JSON files
- ???

sandersn replied:
* plain JS (no check options anywhere)
* checked JS (check turned on locally or globally)
* no-checked JS (check turned **off** locally)
* TS
* no-checked TS (check turned **off** locally)
* external (from plugins?)
* deferred (??)
* ???

I have no idea what's in ???, that's the problem with inverting the predicate.

sandersn commented on src/compiler/program.ts:
```diff
@@ -818,6 +818,25 @@ namespace ts {
+        Diagnostics.Another_export_default_is_here.code,
+        Diagnostics.The_first_export_default_is_here.code,
+        Diagnostics.Identifier_expected_0_is_a_reserved_word_at_the_top_level_of_a_module.code,
+        Diagnostics.Identifier_expected_0_is_a_reserved_word_in_strict_mode_Modules_are_automatically_in_strict_mode.code,
```
double-check that top-level await doesn't get an error in non-modules.

sandersn replied:
Yep, both of the TS errors are custom and are not on this list.

sandersn commented on src/compiler/program.ts:
```diff
@@ -2004,15 +2023,25 @@ namespace ts {
+                const includeBindAndCheckDiagnostics = !isTsNoCheck && (sourceFile.scriptKind === ScriptKind.TS || sourceFile.scriptKind === ScriptKind.TSX
+                        || sourceFile.scriptKind === ScriptKind.External || isPlainJs || isCheckJs || sourceFile.scriptKind === ScriptKind.Deferred);
+                let bindDiagnostics: readonly Diagnostic[] = includeBindAndCheckDiagnostics ? sourceFile.bindDiagnostics : emptyArray;
+                let checkDiagnostics = includeBindAndCheckDiagnostics ? typeChecker.getDiagnostics(sourceFile, cancellationToken) : emptyArray;
```
try out some minified or otherwise compiled JS files to make sure perf doesn't tank or crash.
(Code uses line length to disable certain features)

sandersn commented:
On a desktop with tsserver logging enabled [1], I created a simple project, then opened a new Typescript file and pasted in some code [2].
I opened ajv.min.js, a single-line 121k file, and switched back to my test file. Goto-def was responsive after a second or so.
Then I opened Typescript's own run.js that we use to run tests, a 14 MB file, and switched back to my test file. Goto-def was responsive after 2-3 seconds.

- In both cases, as soon goto-def was responsive, the rest of the language service was responsive.
- Code notified me that it was not tokenising ajv.min.js, and there was no syntax highlighting.
- It took considerably longer to get quick info from run.js itself when I switched back to it. Probably another 10 seconds?

Based on this:

1. I think it's fine to proceed with the simple approach of requesting semantic diagnostics for plain JS files. If a particularly slow use case turns up in the beta period, we'll have a couple of ways to fix it.
2. I'll try to find out why grammar errors are in the checker and see how hard it would be to move them earlier. They make up about 20% of the errors in the checker, and most of the value for plain JS is probably in the grammar errors.

[1] So, faster because it's a desktop; slower because logging is on.
[2] Just some random code I worked on recently.

```ts
export class C {
    set x(...n) { }
}
declare var x: Promise<number>
await x
const y = await x

class Test {
    foo() {

    }

    bar() {
        this.blargh(this, "completely wrong") // Can't autocomplete second parameter

        const t = new Test()
        this.blargh(t, "bar"); // Second parameter autocompletes to "foo", "blargh", "bar"
    }

    blargh<T>(a: T, k: keyof T) {
        // some implementation
    }
}

const t = new Test()
t.blargh(t, "bar")
t.blargh(t, "hi")
```

amcasey commented:
@sandersn How long did it take for GTD to become usable without the diagnostics?  (Sorry if I just missed that in your write-up.)

If we think we might have to roll this back (contextually?), would it be useful to have a configuration option so that editors can add an escape hatch?

sandersn commented:
Good question. After opening ajv.min.js, a lot less than a second. After opening run.js, around 2 seconds -- there probably was a difference from the semantics-requesting version, but not one that I could measure without a stopwatch.

DanielRosenwasser commented:
> If a particularly slow use case turns up in the beta period, we'll have a couple of ways to fix it.

How do you expect most JavaScript users, who don't opt into nightly/beta versions of TypeScript in their editor, to provide feedback during the beta period?

amcasey commented:
I think @andrewbranch looked into ways to monitor completion performance via telemetry when he was working on auto-imports.  Maybe that would help?

sandersn commented:
Now that I think about it, `getSuggestionDiagnostics` is called on plain JS files, and it checks the entire file as well. Calling `getDiagnostics` shouldn't check the file twice, so calling it shouldn't add much time at all.

@mjbvz Can you confirm how much getSuggestionDiagnostics is called on plain JS files compared to getDiagnostics on TS files? I poked around vscode's typescript code but couldn't figure it out. From usage, it seems like suggestions show up on plain JS files at the same rate that errors do on TS files.

mjbvz commented:
@sandersn VS Code uses the `geterr` TS Server command to get all the errors for a file instead of requesting each type of diagnostic individually. Is that what you were asking about?

sandersn commented:
Pretty much--I want to know whether it's calling `geterr` on JS files as often as it is on TS files. I suspect the answer is yes, in order for it to be able to show suggestions in JS files.

mjbvz commented:
Yes we should have the same code path for both JS and TS on the VS Code side
